### PR TITLE
fix: improve feat/transitions type

### PR DESCRIPTION
**Pull request by Talent30**
@franky47  
Re: https://github.com/47ng/next-usequerystate/pull/406#discussion_r1401100668

### CODE REVIEW

**Review by franky47 — CHANGES_REQUESTED** [review]
That's a very interesting approach!

**suggestion**: We could use better descriptive names for generic types than B and S (S stands for Shallow I guess, but is B for Boolean?)

**issue**: The end-to-end test bench fails to compile (run with `pnpm test`):

```
e2e:build: Failed to compile.
e2e:build: 
e2e:build: ./src/pages/pages/useQueryState/index.tsx:40:43
e2e:build: Type error: Type 'false' is not assignable to type 'undefined'.
e2e:build: 
e2e:build:   38 |         <button
e2e:build:   39 |           id="string_set_a"
e2e:build: > 40 |           onClick={() => setString('a', { shallow: false })}
e2e:build:      |                                           ^
e2e:build:   41 |         >
e2e:build:   42 |           Set A
e2e:build:   43 |         </button>
```

**Comment by Talent30** [comment]
> That's a very interesting approach!
> 
> **suggestion**: We could use better descriptive names for generic types than B and S (S stands for Shallow I guess, but is B for Boolean?)
> 
> **issue**: The end-to-end test bench fails to compile (run with `pnpm test`):
> 
> ```
> e2e:build: Failed to compile.
> e2e:build: 
> e2e:build: ./src/pages/pages/useQueryState/index.tsx:40:43
> e2e:build: Type error: Type 'false' is not assignable to type 'undefined'.
> e2e:build: 
> e2e:build:   38 |         <button
> e2e:build:   39 |           id="string_set_a"
> e2e:build: > 40 |           onClick={() => setString('a', { shallow: false })}
> e2e:build:      |                                           ^
> e2e:build:   41 |         >
> e2e:build:   42 |           Set A
> e2e:build:   43 |         </button>
> ```

Any way to get around it in TS? Sorry my brain is not functioning...

**Comment by Talent30** [comment]
It is a crazy way to type it, TBH I don't know how should I name them. Tried my best.

**Comment by franky47** [comment]
Suggested edit:
```diff
diff --git a/packages/next-usequerystate/src/defs.ts b/packages/next-usequerystate/src/defs.ts
index b119442..dde254a 100644
--- a/packages/next-usequerystate/src/defs.ts
+++ b/packages/next-usequerystate/src/defs.ts
@@ -5,8 +5,8 @@ export type Router = ReturnType<typeof useRouter>
 export type HistoryOptions = 'replace' | 'push'
 
 export type Options<
-  AgnosticShallowOption = boolean | unknown,
-  ShallowOption = boolean | undefined
+  Shallow = boolean,
+  StartTransition = Shallow extends true ? never : React.TransitionStartFunction
 > = {
   /**
    * How the query update affects page history
@@ -32,10 +32,7 @@ export type Options<
    * Setting it to `false` will trigger a network request to the server with
    * the updated querystring.
    */
-  shallow?: Extract<
-    ShallowOption extends boolean ? AgnosticShallowOption : boolean,
-    boolean
-  >
+  shallow?: Shallow
 
   /**
    * Maximum amount of time (ms) to wait between updates of the URL query string.
@@ -55,13 +52,7 @@ export type Options<
    *
    * Using this will set the `shallow` setting to `false` automatically.
    */
-  startTransition?: AgnosticShallowOption extends false
-    ? React.TransitionStartFunction
-    : AgnosticShallowOption extends true
-    ? never
-    : AgnosticShallowOption extends unknown
-    ? React.TransitionStartFunction
-    : never
+  startTransition?: StartTransition
 }
 
 export type Nullable<T> = {
diff --git a/packages/next-usequerystate/src/parsers.ts b/packages/next-usequerystate/src/parsers.ts
index dae9382..a4800fe 100644
--- a/packages/next-usequerystate/src/parsers.ts
+++ b/packages/next-usequerystate/src/parsers.ts
@@ -15,10 +15,7 @@ export type ParserBuilder<T> = Required<Parser<T>> &
      * Note that you can override those options in individual calls to the
      * state updater function.
      */
-    withOptions<This, AgnosticShallowOption>(
-      this: This,
-      options: Options<AgnosticShallowOption>
-    ): This
+    withOptions<This>(this: This, options: Options): This
 
     /**
      * Specifying a default value makes the hook state non-nullable when the
diff --git a/packages/next-usequerystate/src/useQueryState.ts b/packages/next-usequerystate/src/useQueryState.ts
index 2d0827c..a384b6a 100644
--- a/packages/next-usequerystate/src/useQueryState.ts
+++ b/packages/next-usequerystate/src/useQueryState.ts
@@ -12,22 +12,20 @@ import {
 } from './update-queue'
 import { safeParse } from './utils'
 
-export interface UseQueryStateOptions<T, AgnosticShallowOption>
-  extends Parser<T>,
-    Options<AgnosticShallowOption> {}
+export interface UseQueryStateOptions<T> extends Parser<T>, Options {}
 
 export type UseQueryStateReturn<Parsed, Default> = [
   Default extends undefined
     ? Parsed | null // value can't be null if default is specified
     : Parsed,
-  <AgnosticShallowOption>(
+  (
     value:
       | null
       | Parsed
       | ((
           old: Default extends Parsed ? Parsed : Parsed | null
         ) => Parsed | null),
-    options?: Options<AgnosticShallowOption>
+    options?: Options
   ) => Promise<URLSearchParams>
 ]
 
@@ -62,9 +60,9 @@ export type UseQueryStateReturn<Parsed, Default> = [
  * @param key The URL query string key to bind to
  * @param options - Parser (defines the state data type), default value and optional history mode.
  */
-export function useQueryState<T, AgnosticShallowOption>(
+export function useQueryState<T>(
   key: string,
-  options: UseQueryStateOptions<T, AgnosticShallowOption> & { defaultValue: T }
+  options: UseQueryStateOptions<T> & { defaultValue: T }
 ): UseQueryStateReturn<
   NonNullable<ReturnType<typeof options.parse>>,
   typeof options.defaultValue
@@ -85,9 +83,9 @@ export function useQueryState<T, AgnosticShallowOption>(
  * @param key The URL query string key to bind to
  * @param options - Parser (defines the state data type), and optional history mode.
  */
-export function useQueryState<T, AgnosticShallowOption>(
+export function useQueryState<T>(
   key: string,
-  options: UseQueryStateOptions<T, AgnosticShallowOption>
+  options: UseQueryStateOptions<T>
 ): UseQueryStateReturn<NonNullable<ReturnType<typeof options.parse>>, undefined>
 
 /**
@@ -121,12 +119,9 @@ export function useQueryState(
  * @param key The URL query string key to bind to
  * @param options - Parser (defines the state data type), and optional history mode.
  */
-export function useQueryState<AgnosticShallowOption>(
+export function useQueryState(
   key: string,
-  options: Pick<
-    UseQueryStateOptions<string, AgnosticShallowOption>,
-    keyof Options
-  >
+  options: Pick<UseQueryStateOptions<string>, keyof Options>
 ): UseQueryStateReturn<string, undefined>
 
 /**
@@ -200,7 +195,7 @@ export function useQueryState(
  * @param key The URL query string key to bind to
  * @param options - Parser (defines the state data type), optional default value and history mode.
  */
-export function useQueryState<T = string, AgnosticShallowOption = boolean>(
+export function useQueryState<T = string>(
   key: string,
   {
     history = 'replace',
@@ -211,7 +206,7 @@ export function useQueryState<T = string, AgnosticShallowOption = boolean>(
     serialize = String,
     defaultValue = undefined,
     startTransition
-  }: Partial<UseQueryStateOptions<T, AgnosticShallowOption>> & {
+  }: Partial<UseQueryStateOptions<T>> & {
     defaultValue?: T
   } = {
     history: 'replace',
diff --git a/packages/next-usequerystate/src/useQueryStates.ts b/packages/next-usequerystate/src/useQueryStates.ts
index 6fdebc8..1aa2d9a 100644
--- a/packages/next-usequerystate/src/useQueryStates.ts
+++ b/packages/next-usequerystate/src/useQueryStates.ts
@@ -24,8 +24,7 @@ export type UseQueryStatesKeysMap<Map = any> = {
   [Key in keyof Map]: KeyMapValue<Map[Key]>
 }
 
-export interface UseQueryStatesOptions<AgnosticShallowOption>
-  extends Options<AgnosticShallowOption> {}
+export interface UseQueryStatesOptions extends Options {}
 
 export type Values<T extends UseQueryStatesKeysMap> = {
   [K in keyof T]: T[K]['defaultValue'] extends NonNullable<
@@ -65,7 +64,7 @@ export function useQueryStates<KeyMap extends UseQueryStatesKeysMap>(
     shallow = true,
     throttleMs = FLUSH_RATE_LIMIT_MS,
     startTransition
-  }: Partial<UseQueryStatesOptions<boolean>> = {}
+  }: Partial<UseQueryStatesOptions> = {}
 ): UseQueryStatesReturn<KeyMap> {
   type V = Values<KeyMap>
   const keys = Object.keys(keyMap).join(',')

```

**Comment by Talent30** [comment]
> 我尝试了一种不同的方法，将类型排除逻辑移至泛型定义中。 它仍然没有通过整个测试套件（它允许使用shallow: true + startTransition，但它不应该这样做），但其他分配是正确的。
> 
> 我开始怀疑这是否可以正确输入。
> 建议补丁
> 
> ```diff
> diff --git a/packages/next-usequerystate/src/defs.ts b/packages/next-usequerystate/src/defs.ts
> index b119442..dde254a 100644
> --- a/packages/next-usequerystate/src/defs.ts
> +++ b/packages/next-usequerystate/src/defs.ts
> @@ -5,8 +5,8 @@ export type Router = ReturnType<typeof useRouter>
>  export type HistoryOptions = 'replace' | 'push'
>  
>  export type Options<
> -  AgnosticShallowOption = boolean | unknown,
> -  ShallowOption = boolean | undefined
> +  Shallow = boolean,
> +  StartTransition = Shallow extends true ? never : React.TransitionStartFunction
>  > = {
>    /**
>     * How the query update affects page history
> @@ -32,10 +32,7 @@ export type Options<
>     * Setting it to `false` will trigger a network request to the server with
>     * the updated querystring.
>     */
> -  shallow?: Extract<
> -    ShallowOption extends boolean ? AgnosticShallowOption : boolean,
> -    boolean
> -  >
> +  shallow?: Shallow
>  
>    /**
>     * Maximum amount of time (ms) to wait between updates of the URL query string.
> @@ -55,13 +52,7 @@ export type Options<
>     *
>     * Using this will set the `shallow` setting to `false` automatically.
>     */
> -  startTransition?: AgnosticShallowOption extends false
> -    ? React.TransitionStartFunction
> -    : AgnosticShallowOption extends true
> -    ? never
> -    : AgnosticShallowOption extends unknown
> -    ? React.TransitionStartFunction
> -    : never
> +  startTransition?: StartTransition
>  }
>  
>  export type Nullable<T> = {
> diff --git a/packages/next-usequerystate/src/parsers.ts b/packages/next-usequerystate/src/parsers.ts
> index dae9382..a4800fe 100644
> --- a/packages/next-usequerystate/src/parsers.ts
> +++ b/packages/next-usequerystate/src/parsers.ts
> @@ -15,10 +15,7 @@ export type ParserBuilder<T> = Required<Parser<T>> &
>       * Note that you can override those options in individual calls to the
>       * state updater function.
>       */
> -    withOptions<This, AgnosticShallowOption>(
> -      this: This,
> -      options: Options<AgnosticShallowOption>
> -    ): This
> +    withOptions<This>(this: This, options: Options): This
>  
>      /**
>       * Specifying a default value makes the hook state non-nullable when the
> diff --git a/packages/next-usequerystate/src/useQueryState.ts b/packages/next-usequerystate/src/useQueryState.ts
> index 2d0827c..a384b6a 100644
> --- a/packages/next-usequerystate/src/useQueryState.ts
> +++ b/packages/next-usequerystate/src/useQueryState.ts
> @@ -12,22 +12,20 @@ import {
>  } from './update-queue'
>  import { safeParse } from './utils'
>  
> -export interface UseQueryStateOptions<T, AgnosticShallowOption>
> -  extends Parser<T>,
> -    Options<AgnosticShallowOption> {}
> +export interface UseQueryStateOptions<T> extends Parser<T>, Options {}
>  
>  export type UseQueryStateReturn<Parsed, Default> = [
>    Default extends undefined
>      ? Parsed | null // value can't be null if default is specified
>      : Parsed,
> -  <AgnosticShallowOption>(
> +  (
>      value:
>        | null
>        | Parsed
>        | ((
>            old: Default extends Parsed ? Parsed : Parsed | null
>          ) => Parsed | null),
> -    options?: Options<AgnosticShallowOption>
> +    options?: Options
>    ) => Promise<URLSearchParams>
>  ]
>  
> @@ -62,9 +60,9 @@ export type UseQueryStateReturn<Parsed, Default> = [
>   * @param key The URL query string key to bind to
>   * @param options - Parser (defines the state data type), default value and optional history mode.
>   */
> -export function useQueryState<T, AgnosticShallowOption>(
> +export function useQueryState<T>(
>    key: string,
> -  options: UseQueryStateOptions<T, AgnosticShallowOption> & { defaultValue: T }
> +  options: UseQueryStateOptions<T> & { defaultValue: T }
>  ): UseQueryStateReturn<
>    NonNullable<ReturnType<typeof options.parse>>,
>    typeof options.defaultValue
> @@ -85,9 +83,9 @@ export function useQueryState<T, AgnosticShallowOption>(
>   * @param key The URL query string key to bind to
>   * @param options - Parser (defines the state data type), and optional history mode.
>   */
> -export function useQueryState<T, AgnosticShallowOption>(
> +export function useQueryState<T>(
>    key: string,
> -  options: UseQueryStateOptions<T, AgnosticShallowOption>
> +  options: UseQueryStateOptions<T>
>  ): UseQueryStateReturn<NonNullable<ReturnType<typeof options.parse>>, undefined>
>  
>  /**
> @@ -121,12 +119,9 @@ export function useQueryState(
>   * @param key The URL query string key to bind to
>   * @param options - Parser (defines the state data type), and optional history mode.
>   */
> -export function useQueryState<AgnosticShallowOption>(
> +export function useQueryState(
>    key: string,
> -  options: Pick<
> -    UseQueryStateOptions<string, AgnosticShallowOption>,
> -    keyof Options
> -  >
> +  options: Pick<UseQueryStateOptions<string>, keyof Options>
>  ): UseQueryStateReturn<string, undefined>
>  
>  /**
> @@ -200,7 +195,7 @@ export function useQueryState(
>   * @param key The URL query string key to bind to
>   * @param options - Parser (defines the state data type), optional default value and history mode.
>   */
> -export function useQueryState<T = string, AgnosticShallowOption = boolean>(
> +export function useQueryState<T = string>(
>    key: string,
>    {
>      history = 'replace',
> @@ -211,7 +206,7 @@ export function useQueryState<T = string, AgnosticShallowOption = boolean>(
>      serialize = String,
>      defaultValue = undefined,
>      startTransition
> -  }: Partial<UseQueryStateOptions<T, AgnosticShallowOption>> & {
> +  }: Partial<UseQueryStateOptions<T>> & {
>      defaultValue?: T
>    } = {
>      history: 'replace',
> diff --git a/packages/next-usequerystate/src/useQueryStates.ts b/packages/next-usequerystate/src/useQueryStates.ts
> index 6fdebc8..1aa2d9a 100644
> --- a/packages/next-usequerystate/src/useQueryStates.ts
> +++ b/packages/next-usequerystate/src/useQueryStates.ts
> @@ -24,8 +24,7 @@ export type UseQueryStatesKeysMap<Map = any> = {
>    [Key in keyof Map]: KeyMapValue<Map[Key]>
>  }
>  
> -export interface UseQueryStatesOptions<AgnosticShallowOption>
> -  extends Options<AgnosticShallowOption> {}
> +export interface UseQueryStatesOptions extends Options {}
>  
>  export type Values<T extends UseQueryStatesKeysMap> = {
>    [K in keyof T]: T[K]['defaultValue'] extends NonNullable<
> @@ -65,7 +64,7 @@ export function useQueryStates<KeyMap extends UseQueryStatesKeysMap>(
>      shallow = true,
>      throttleMs = FLUSH_RATE_LIMIT_MS,
>      startTransition
> -  }: Partial<UseQueryStatesOptions<boolean>> = {}
> +  }: Partial<UseQueryStatesOptions> = {}
>  ): UseQueryStatesReturn<KeyMap> {
>    type V = Values<KeyMap>
>    const keys = Object.keys(keyMap).join(',')
> ```

I don't think it is going to be feasible.

File: packages/next-usequerystate/src/tests/parsers.test-d.ts [hunk 1]
```diff
@@ -27,3 +27,36 @@ import { parseAsString } from '../../dist'
...
+// Shallow / startTransition interaction
+{
+  expectType<boolean | undefined>(parseAsString.withOptions({}).shallow)
+  expectType<boolean | undefined>(
+    parseAsString.withOptions({ shallow: true }).shallow
+  )
+  expectType<boolean | undefined>(
+    parseAsString.withOptions({ shallow: false }).shallow
+  )
+  // Should probably be `false` here, but it's not worth the complexity
```
Comment (Talent30) [inline packages/next-usequerystate/src/tests/parsers.test-d.ts:40]: Agree, otherwise the types are going to be a circus...

**Comment by Talent30** [comment]
@franky47 This is the best I can do. Please approve this PR if you are happy with it.

File: packages/next-usequerystate/src/tests/parsers.test-d.ts (hunk 1 above)
Comment (franky47) [reply packages/next-usequerystate/src/tests/parsers.test-d.ts:40]: Actually, `boolean | undefined` is valid here, as `withOptions` doesn't set `shallow` when setting `startTransition`, it's done when merging options in the URL update queue.

**Review by franky47 — APPROVED** [review]
**praise**: Great job! 👏 

There are many corner cases where we could end up, like chaining `withOptions`, or overriding hook-level options with call-level ones, but I realised that the type for the Options object still make sense to be `boolean | undefined` and `React.TransitionStartFunction | undefined`, as the setting of `shallow: false` when passing a `startTransition` is not done when defining options, but when they get merged when applied to the URL queue.

**Pull request merged**